reference-implementation/lib/readable-byte-stream.js

```diff
@@ -0,0 +1,948 @@
+    if (this._state === 'closed') {
+      return Promise.resolve(CreateIterResultObject(undefined, true));
+    }
+
+    if (this._state === 'errored') {
+      return Promise.reject(this._storedError);
+    }
+
+    assert(this._ownerReadableByteStream !== undefined, 'This reader must be attached to a stream');
+    assert(this._ownerReadableByteStream._state === 'readable', 'The owner stream must be in readable state');
+
+    const promise = new Promise((resolve, reject) => {
+      this._readRequests.push({resolve, reject});
+    });
+
+    PullFromReadableByteStream(this._ownerReadableByteStream);
+
+    return promise;
+  }
+
+  releaseLock() {
+    if (!IsReadableByteStreamReader(this)) {
+      throw new TypeError(
+          'ReadableByteStreamReader.prototype.releaseLock can only be used on a ReadableByteStreamReader');
+    }
+
+    if (this._ownerReadableByteStream === undefined) {
+      return;
+    }
+
+    if (this._readRequests.length > 0) {
+      throw new TypeError('Tried to release a reader lock when that reader has pending read() calls un-settled');
+    }
+
+    if (this._ownerReadableByteStream._state === 'errored') {
+      MarkReadableByteStreamReaderErrored(this, this._ownerReadableByteStream._storedError)
+    } else {
+      CloseReadableByteStreamReader(this);
+    }
+  }
+}
+
+class ReadableByteStreamByobReader {
+  constructor(stream) {
+    if (!IsReadableByteStream(stream)) {
+      throw new TypeError('ReadableByteStreamByobReader can only be constructed with a ReadableByteStream instance');
+    }
+    if (IsReadableByteStreamLocked(stream)) {
+      throw new TypeError('This stream has already been locked for exclusive reading by another reader');
+    }
+
+    InitializeReadableByteStreamReader(this, stream);
+
+    this._readIntoRequests = [];
+  }
+
+  get closed() {
+    if (!IsReadableByteStreamByobReader(this)) {
+      return Promise.reject(
+        new TypeError(
+            'ReadableByteStreamByobReader.prototype.closed can only be used on a ReadableByteStreamByobReader'));
+    }
+
+    return this._closedPromise;
+  }
+
+  cancel(reason) {
+    if (!IsReadableByteStreamByobReader(this)) {
+      return Promise.reject(
+        new TypeError(
+            'ReadableByteStreamByobReader.prototype.cancel can only be used on a ReadableByteStreamByobReader'));
+    }
+
+    if (this._state === 'closed') {
+      return Promise.resolve(undefined);
+    }
+
+    if (this._state === 'errored') {
+      return Promise.reject(this._storedError);
+    }
+
+    assert(this._ownerReadableByteStream !== undefined, 'This stream must be attached to a stream');
+
+    return CancelReadableByteStream(this._ownerReadableByteStream, reason);
+  }
+
+  read(view) {
+    if (!IsReadableByteStreamByobReader(this)) {
+      return Promise.reject(
+        new TypeError(
+            'ReadableByteStreamByobReader.prototype.read can only be used on a ReadableByteStreamByobReader'));
+    }
+
+    if (view === undefined) {
```

Comment:
I realize the above may be confusing since I am talking mostly about how to write the spec while commenting on the code. For the code, I think just use `ArrayBuffer.isView` here, and then continue using `.byteLength` etc. as below.